pkg/rpc/context.go

```diff
@@ -2042,32 +2060,91 @@ func (rpcCtx *Context) GRPCDialPod(
 	return rpcCtx.GRPCDialNode(target, roachpb.NodeID(remoteInstanceID), class)
 }
 
+type Peer struct {
+	c *Connection // mutate only when surrounding map's mutex is held exclusively
+
+	mu struct {
+		syncutil.RWMutex
+		// disconnected is zero initially, reset on successful heartbeat, set on
+		// heartbeat teardown if zero. In other words, does not move forward across
+		// subsequent connection failures - it tracks the first disconnect since
+		// having been healthy.
+		disconnected time.Time
+	}
+}
+
+func (p *Peer) conn() (*Connection, bool) {
+	if p == nil {
+		return nil, false
+	}
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	if !p.mu.disconnected.IsZero() {
+		return nil, false
+	}
+	return p.c, p.c != nil
+}
+
 type connMap struct {
 	mu struct {
 		syncutil.RWMutex
-		m map[connKey]*Connection
+		// TODO(koorosh): now that we hold on to connections forever, add a watcher
+		// goroutine to rpc.Context that will periodically try to reconnect to peers
+		// which have a nil `*Connection`.
+		//
+		// The tricky part is when to evict from this map. We should do so once the
+		// peer is known to have been fully decommissioned. This can be done via a
+		// gossip callback on the liveness. This will then also easily address [1]
+		// since that callback can make sure that `m` contains an entry for each
+		// node which is a member of the cluster (assuming we have a full view of
+		// gossip but if we don't have that we also can't do much better by other
+		// means).
+		//
+		// [1]: https://github.com/cockroachdb/cockroach/issues/70111
+		m map[connKey]*Peer
 	}
 }
 
 func (m *connMap) Get(k connKey) (*Connection, bool) {
 	m.mu.RLock()
 	defer m.mu.RUnlock()
-	c, ok := m.mu.m[k]
-	return c, ok
+	pc := m.mu.m[k]
+	return pc.conn()
 }
 
-func (m *connMap) Remove(k connKey, conn *Connection) error {
+func (m *connMap) GetPeer(k connKey) (*Peer, bool) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	pc, ok := m.mu.m[k]
+	return pc, ok
+}
+
+// RemoveNodeConnections removes all connections to target node. It is used when some node is
+// decommissioned and removed from the cluster.
+func (m *connMap) RemoveNodeConnections(nodeID roachpb.NodeID) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for c := range m.mu.m {
+		if c.nodeID == nodeID {
+			delete(m.mu.m, c)
```

Comment:
Done. `decommissioned` flag is used instead of removing records from map.
Now, when connection fails, we derive information whether node is decommission based on error type that in turn based on `nodeTombstoneStorage` logic.